modyn/supervisor/internal/triggers/datadrifttrigger.py

```diff
@@ -0,0 +1,260 @@
+            (x["result"]["drift_score"], x["result"]["method_name"], x["result"]["drift_detected"])
+            for x in result["metrics"]
+        ]
+        logger.info(
+            f"[DataDriftDetector][Prev Trigger {self.previous_trigger_id}][Dataset {self.dataloader_info.dataset_id}]"
+            + f"[Result] {result_print}"
+        )
+
+        with open(self.drift_dir / f"drift_{self.previous_trigger_id}.json", "w", encoding="utf-8") as f:
+            json.dump(result, f)
+
+        return result["metrics"][0]["result"]["drift_detected"]
+
+    def detect_drift(self, idx_start: int, idx_end: int) -> bool:
+        assert self.previous_trigger_id is not None
+        assert self.previous_data_points is not None and self.previous_data_points > 0
+        assert self.previous_model_id is not None
+        assert self.dataloader_info is not None
+        assert self.model_downloader is not None
+
+        reference_dataloader = prepare_trigger_dataloader_by_trigger(
+            self.previous_trigger_id,
+            self.dataloader_info,
+            data_points_in_trigger=self.previous_data_points,
+            sample_size=self.sample_size,
+        )
+
+        current_keys, _, _ = zip(*self.data_cache[idx_start:idx_end])  # type: ignore
+        current_dataloader = prepare_trigger_dataloader_fixed_keys(
+            self.previous_trigger_id + 1,
+            self.dataloader_info,
+            current_keys,
+            sample_size=self.sample_size,
+        )
+
+        # Download previous model as embedding encoder
+        # TODO(366) Support custom model as embedding encoder
+        if self.model_updated:
+            self.model_downloader.download(self.previous_model_id)
+            self.model_updated = False
+
+        # Compute embeddings
+        reference_embeddings_df = get_embeddings_evidently_format(self.model_downloader, reference_dataloader)
+        current_embeddings_df = get_embeddings_evidently_format(self.model_downloader, current_dataloader)
+
+        drift_detected = self.run_detection(reference_embeddings_df, current_embeddings_df)
+
+        return drift_detected
+
+    def inform(self, new_data: list[tuple[int, int, int]]) -> Generator[int, None, None]:
+        """
+        Use Generator here because this data drift trigger
+        needs to wait for the previous trigger to finish and get the model
+        """
+        # add new data to data_cache
+        self.data_cache.extend(new_data)
+
+        unvisited_data_points = len(self.data_cache)
+        untriggered_data_points = unvisited_data_points
+        # the sliding window of data points for detection
+        detection_idx_start = 0
+        detection_idx_end = 0
+
+        while unvisited_data_points >= self.detection_interval:
+            unvisited_data_points -= self.detection_interval
+            detection_idx_end += self.detection_interval
+            if detection_idx_end <= self.leftover_data_points:
+                continue
```

Comment:
Should we move this at the end of the loop? Semenatically, we have not visited any new points here, have we? (Generally, I think some description of the algorithm as mentioned in the docstring comment would be super helpful)

Reply:
It must be handled at the top because it is including all the remaining untriggered samples from the previous new data batch. The remaining untriggered samples have been processed in the previous batch, so there's no need to process them again. Explained in the docstring.